Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/cache/stream/WebmStreamingSource.kt

```diff
@@ -110,6 +96,164 @@ class WebmStreamingSource(
+        Logger.d(TAG, "createMediaSource() Loading already downloaded file from the disk")
+        val fileUri = Uri.parse(rawFile.getFullPath())
+
+        callback.onMediaSourceReady(
+                ProgressiveMediaSource.Factory(DataSource.Factory { FileDataSource() })
+                        .createMediaSource(fileUri)
+        )
+    }
+
+    private fun loadFromLocalThread(
+            postImage: PostImage,
+            loadable: Loadable,
+            callback: MediaSourceCallback
+    ) {
+        Logger.d(TAG, "createMediaSource() Loading local thread file")
+
+        val filename = ThreadSaveManager.formatOriginalImageName(
+                postImage.serverFilename,
+                postImage.extension
+        )
+
+        val imagesSubDirSegments = ThreadSaveManager.getImagesSubDir(loadable)
+        val segments: MutableList<Segment> = ArrayList(imagesSubDirSegments).apply {
+            add(FileSegment(filename))
+        }
+
+        val baseDirFile = fileManager.newBaseDirectoryFile(
+                LocalThreadsBaseDirectory::class.java
+        )
+
+        if (baseDirFile == null) {
+            BackgroundUtils.runOnMainThread {
+                callback.onError(
+                        IOException("createMediaSource() Couldn't create a file " +
+                                "inside local threads base directory")
+                )
+            }
+
+            return
+        }
+
+        val localImgFile = baseDirFile.clone(segments)
+        val exists = fileManager.exists(localImgFile)
+        val isFile = fileManager.isFile(localImgFile)
+        val canRead = fileManager.canRead(localImgFile)
+        val isLocalFileOk = exists && isFile && canRead
+
+        if (!isLocalFileOk) {
+            BackgroundUtils.runOnMainThread {
+                callback.onError(
+                        IOException("createMediaSource() Couldn't open webm from a local thread " +
+                                "(exists = $exists, isFile = $isFile, canRead = $canRead)")
+                )
+            }
+
+            return
+        }
+
+        // FileDataSource uses RandomAccessFile internally, which, apparently cannot read files
+        // stored via SAF, so we have to copy the file into the cache directory to use it
+        val videoUrl = postImage.imageUrl.toString()
+        val destination = createDestinationFile(videoUrl, localImgFile, callback)
+        if (destination == null) {
+            // No need to call callbacks it's already called in createDestinationFile
+            return
+        }
+
+        val fileUri = Uri.parse(destination.getFullPath())
+
+        BackgroundUtils.runOnMainThread {
+            callback.onMediaSourceReady(
+                    ProgressiveMediaSource.Factory(DataSource.Factory { FileDataSource() })
+                            .createMediaSource(fileUri)
+            )
+        }
+    }
+
+    private fun createDestinationFile(
```

Comment:
This function is remarkably similar to FileCacheV2's handleLocalThreadFileImmediatelyAvailable, maybe the two can be combined into one?

Reply:
Yeah, they are pretty much the same. Good catch!